src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs

```diff
@@ -307,30 +308,42 @@ public void GarbageCollectInstalledWorkloadPacks(DirectoryPath? offlineCache = n
             {
                 foreach (var packVersionDir in Directory.GetDirectories(packIdDir))
                 {
-                    var bandRecords = Directory.GetFileSystemEntries(packVersionDir);
+                    var bandRecordPaths = Directory.GetFileSystemEntries(packVersionDir);
 
-                    var unneededBandRecords = bandRecords
-                        .Where(recordPath => !installedSdkFeatureBands.Contains(new SdkFeatureBand(Path.GetFileName(recordPath))));
+                    // Mark packs for GC from versions / feature bands of the SDK that are not present on the machine and lower or equal to the current sdk.
+                    var unneededBandRecordPaths = bandRecordPaths
+                        .Where(recordPath =>
+                            {
+                                var featureBand = new SdkFeatureBand(Path.GetFileName(recordPath));
+                                return featureBand.CompareTo(_sdkFeatureBand) < 1 &&
+                                    (!installedSdkFeatureBands.Contains(featureBand) || cleanAllPacks );
+                            }
+                    );
 
+                    // Mark packs for GC that don't have a corresponding workload installation record (or all packs).
                     var currentBandRecordPath = Path.Combine(packVersionDir, _sdkFeatureBand.ToString());
-                    if (bandRecords.Contains(currentBandRecordPath) && !currentBandInstallRecords.Contains(currentBandRecordPath))
+                    if (
+                        bandRecordPaths.Contains(currentBandRecordPath) && // A pack for this band exists.
+                        new SdkFeatureBand(Path.GetFileName(currentBandRecordPath)).CompareTo(_sdkFeatureBand) < 1 && // We can parse the format of that band (its older than our band or current.)
```

Comment:
> CompareTo results should be compared against zero.

I am confused. `band.CompareTo(band) == 0`. It is meant to be if the feature band is equal to or less than. If it is < 0 then it is only doing less than the current band. Do you mean `<=`? 

But sounds good. Let's just do it for every band.

Reply:
You're right, it should be `<= 0`, and since `CompareTo` returns an integer, `<= 0` is the same as `< 1`.  I was thinking the value might be a `double`.